In [1]:
from keras.models import Sequential
from keras.layers import Dropout
from keras.layers import Dense, Activation,Dropout
from keras import regularizers
from keras.layers.advanced_activations import LeakyReLU
from keras.layers.normalization import BatchNormalization

import numpy as np
# fix random seed for reproducibility
np.random.seed(42)

import pandas as pd

Using TensorFlow backend.


In [2]:
#Folder for the dataset
datasetFolder = '/home/carnd/dbpedia2016/inRel_1xall/dataset/'

#Number of files
numberOfFiles = 638

#Test split
testSplit=0.2

In [3]:
def load_data(datasetFolder, datasetXFile, datasetYFile, wrap=False, printIt=False):
    #print('Loading X')
    # load file
    with open(datasetFolder + datasetXFile, "r") as f:
        head = f.readline()
        cols = head.split(',')
        numberOfCols = len(cols)
        #print(numberOfCols)
        numberOfRows=0
        for line in f:
            numberOfRows+=1
        f.close()
    if(printIt):
        print('Input Features: {} x {}'.format(numberOfRows,numberOfCols))
    if(wrap==True):
        maxY = 1000
        shift = (1+((numberOfCols-1)//maxY))*0.5
    else:
        maxY = numberOfCols-1
        shift = 0.5
        
    dataX = np.zeros([numberOfRows,maxY],np.float16)
    with open(datasetFolder + datasetXFile, "r") as f:
        head = f.readline()
        rowCounter=0
        for line in f:
            row=line.split(',')
            for i in range(1, len(row)):
                if(int(row[i])<=0):
                    continue;
                dataX[rowCounter][(int(row[i])-1)%maxY]=1 + ((int(row[i])-1)//maxY) - shift
                #if((1 + ((int(row[i])-1)//maxY))>1):
                #    print("{} data[{}][{}] = {}".format(int(row[i])-1, rowCounter,(int(row[i])-1)%maxY,1 + ((int(row[i])-1)//maxY)))
            rowCounter+=1
        f.close()
   
    #print('Loading Y')
    # load file
    with open(datasetFolder + datasetYFile, "r") as f:
        head = f.readline()
        cols = head.split(',')
        numberOfCols = len(cols)
        #print(numberOfCols)
        numberOfRows=0
        for line in f:
            numberOfRows+=1
        f.close()

    if(printIt):
        print('Output Features: {} x {}'.format(numberOfRows,numberOfCols))
    dataY = np.zeros([numberOfRows,(numberOfCols-1)],np.float16)
    with open(datasetFolder + datasetYFile, "r") as f:
        head = f.readline()
        rowCounter=0
        for line in f:
            row=line.split(',')
            for i in range(1, len(row)):
                if(int(row[i])<=0):
                    continue;
                dataY[rowCounter][(int(row[i])-1)]=1
            rowCounter+=1
        f.close()
        

    return dataX, dataY

In [4]:
dataX, dataY = load_data(datasetFolder,'datasetX_1.csv', 'datasetY_1.csv', printIt=True)

Input Features: 1109 x 1075
Output Features: 1109 x 526


In [5]:
dataX, dataY = load_data(datasetFolder,'datasetX_1.csv', 'datasetY_1.csv')

In [6]:
print(dataX.shape)
print(dataX[0:5])

(1109, 1074)
[[ 0.5  0.   0.  ...,  0.   0.   0. ]
 [ 0.   0.5  0.5 ...,  0.   0.   0. ]
 [ 0.   0.   0.  ...,  0.   0.   0. ]
 [ 0.   0.   0.  ...,  0.   0.   0. ]
 [ 0.   0.   0.  ...,  0.   0.   0. ]]


In [7]:
print(dataY.shape)
print(dataY[0:5])

(1109, 525)
[[ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]
 [ 0.  0.  0. ...,  0.  0.  0.]]


In [8]:
print("Input Features for classification: {}".format(dataX.shape[1]))
print("Output Classes for classification: {}".format(dataY.shape[1]))

Input Features for classification: 1074
Output Classes for classification: 525


In [9]:
logisticRegression = Sequential(name='Simple Logistic Regression')
logisticRegression.add(Dense(dataY.shape[1], input_dim=dataX.shape[1], activation='sigmoid', init='glorot_uniform'))

deepModel = Sequential(name='Deep Model (5 Dense Layers)')
deepModel.add(Dense(1048, input_dim=dataX.shape[1], init='glorot_normal'))
deepModel.add(BatchNormalization())
deepModel.add(Activation('relu'))
deepModel.add(Dropout(0.2))
deepModel.add(Dense(1048, init='glorot_normal'))
deepModel.add(BatchNormalization())
deepModel.add(Activation('relu'))
deepModel.add(Dropout(0.2))
deepModel.add(Dense(1048, init='glorot_normal'))
deepModel.add(BatchNormalization())
deepModel.add(Activation('relu'))
deepModel.add(Dropout(0.2))
deepModel.add(Dense(dataY.shape[1], activation='sigmoid', init='glorot_normal'))

reallyDeepModel = Sequential(name='Deep Model (8 Dense Layers)')
reallyDeepModel.add(Dense(512, input_dim=dataX.shape[1], init='glorot_normal'))
reallyDeepModel.add(BatchNormalization())
reallyDeepModel.add(Activation('relu'))
reallyDeepModel.add(Dropout(0.2))
reallyDeepModel.add(Dense(512, input_dim=dataX.shape[1], init='glorot_normal'))
reallyDeepModel.add(BatchNormalization())
reallyDeepModel.add(Activation('relu'))
reallyDeepModel.add(Dropout(0.2))
reallyDeepModel.add(Dense(512, input_dim=dataX.shape[1], init='glorot_normal'))
reallyDeepModel.add(BatchNormalization())
reallyDeepModel.add(Activation('relu'))
reallyDeepModel.add(Dropout(0.2))
reallyDeepModel.add(Dense(512, input_dim=dataX.shape[1], init='glorot_normal'))
reallyDeepModel.add(BatchNormalization())
reallyDeepModel.add(Activation('relu'))
reallyDeepModel.add(Dropout(0.2))
reallyDeepModel.add(Dense(512, input_dim=dataX.shape[1], init='glorot_normal'))
reallyDeepModel.add(BatchNormalization())
reallyDeepModel.add(Activation('relu'))
reallyDeepModel.add(Dropout(0.2))
reallyDeepModel.add(Dense(512, input_dim=dataX.shape[1], init='glorot_normal'))
reallyDeepModel.add(BatchNormalization())
reallyDeepModel.add(Activation('relu'))
reallyDeepModel.add(Dropout(0.2))
reallyDeepModel.add(Dense(512, input_dim=dataX.shape[1], init='glorot_normal'))
reallyDeepModel.add(BatchNormalization())
reallyDeepModel.add(Activation('relu'))
reallyDeepModel.add(Dropout(0.2))
reallyDeepModel.add(Dense(dataY.shape[1], activation='sigmoid', init='glorot_normal'))

models = [logisticRegression, deepModel, reallyDeepModel]

In [10]:
# Compile model
import keras.backend as K

def count_predictions(y_true, y_pred):
    true_positives = K.sum(K.round(K.clip(y_true * y_pred, 0, 1)))
    predicted_positives = K.sum(K.round(K.clip(y_pred, 0, 1)))
    possible_positives = K.sum(K.round(K.clip(y_true, 0, 1)))
    return true_positives, predicted_positives, possible_positives

def f1score(y_true, y_pred):
    true_positives, predicted_positives, possible_positives = count_predictions(y_true, y_pred)
    precision = true_positives / (predicted_positives + K.epsilon())
    recall = true_positives / (possible_positives + K.epsilon())
    f1score = 2.0 * precision * recall / (precision+recall+ K.epsilon())
    return f1score

def fBetaScore(y_true, y_pred, beta):
    true_positives, predicted_positives, possible_positives = count_predictions(y_true, y_pred)
    precision = true_positives / (predicted_positives + K.epsilon())
    recall = true_positives / (possible_positives + K.epsilon())
    f1score = (1+(beta*beta)) * precision * recall / ((beta*beta*precision)+recall+ K.epsilon())
    return f1score

for model in models:
    model.compile(loss='binary_crossentropy', optimizer='nadam', metrics=[f1score])

In [11]:
def fit_data(model, dataX, dataY):
    # Fit the model
    #model.fit(dataX, dataY, nb_epoch=5, verbose=2, batch_size=256)
    return model.train_on_batch(dataX, dataY)

In [12]:
def countPredictions(y_true, y_pred):
    true_positives = np.sum(np.round(y_pred*y_true))
    predicted_positives = np.sum(np.round(y_pred))
    possible_positives = np.sum(y_true)
    return true_positives, predicted_positives, possible_positives

In [15]:
#Randomize the list of numbers so we can split train and test dataset
listOfFiles=list(range(1,numberOfFiles+1))
import random
random.shuffle(listOfFiles)
splitIndex=int((1-testSplit)*numberOfFiles)

numberOfEons = 10
for eon in range(0, numberOfEons):
    print('{}. Eon {}/{}'.format(eon+1,eon+1, numberOfEons))
    for trainIndex in range(0,splitIndex):
        dataX, dataY = load_data(datasetFolder,'datasetX_{}.csv'.format(listOfFiles[trainIndex]), 'datasetY_{}.csv'.format(listOfFiles[trainIndex]))
        for model in models:
            #print('Model = {}'.format(model.name))
            model.fit(dataX, dataY, nb_epoch=1, verbose=0, batch_size=512)
            sc=model.test_on_batch(dataX,dataY)
            loss = sc[0]
            f1score = sc[1]
            #loss, f1score=fit_data(model,dataX, dataY)
            print('Learning for file {} / {} : datasetX/Y_{}\t\tloss={:.4f} f1score={:.4f}'.format(trainIndex+1, splitIndex, listOfFiles[trainIndex], loss, f1score), end='\r')
        
    counts = {} 
    for model in models:
        counts[model.name] = {'true_positives':0, 'predicted_positives':0, 'possible_positives':0}
    
    for testIndex in range(splitIndex, numberOfFiles):
        dataX, dataY = load_data(datasetFolder,'datasetX_{}.csv'.format(listOfFiles[testIndex]), 'datasetY_{}.csv'.format(listOfFiles[testIndex]))
        for model in models:
            predY=model.predict_on_batch(dataX)
            true_positives, predicted_positives, possible_positives = countPredictions(dataY, predY)
            counts[model.name]['true_positives'] += true_positives
            counts[model.name]['predicted_positives'] += predicted_positives
            counts[model.name]['possible_positives'] += possible_positives
            print ('Testing for file {} / {} : datasetX/Y_{} - true +ve:{}  pred +ve:{} possible +ve:{}'.format(testIndex+1, numberOfFiles, listOfFiles[testIndex], true_positives,predicted_positives,possible_positives), end='\r')
    
    for model in models:
        count = counts[model.name]
        precision = (count['true_positives'])/(count['predicted_positives']+0.0001)
        recall = (count['true_positives'])/(count['possible_positives']+0.0001)
        f1score = 2.0 * precision * recall / (precision+recall+0.0001)
        print(' - Model = {} \t f1-score = {:.4f}\t precision = {:.4f} \t recall = {:.4f}'.format(model.name, f1score, precision, recall))

1. Eon 1/10
 - Model = Simple Logistic Regression 	 f1-score = 0.5093	 precision = 0.8421 	 recall = 0.36502.0
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.7295	 precision = 0.7439 	 recall = 0.7158
 - Model = Deep Model (8 Dense Layers) 	 f1-score = 0.7538	 precision = 0.7647 	 recall = 0.7433
2. Eon 2/10
 - Model = Simple Logistic Regression 	 f1-score = 0.5780	 precision = 0.8399 	 recall = 0.44062.0
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.7526	 precision = 0.7822 	 recall = 0.7252
 - Model = Deep Model (8 Dense Layers) 	 f1-score = 0.7432	 precision = 0.7884 	 recall = 0.7030
3. Eon 3/10
 - Model = Simple Logistic Regression 	 f1-score = 0.6246	 precision = 0.8295 	 recall = 0.50092.0
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.7354	 precision = 0.8096 	 recall = 0.6737
 - Model = Deep Model (8 Dense Layers) 	 f1-score = 0.7558	 precision = 0.7734 	 recall = 0.7392
4. Eon 4/10
 - Model = Simple Logistic Regression 	 f1-score = 0.6462	 precision = 0.

KeyboardInterrupt: 

==================================================
# inRel_1xall
==================================================
1. Eon 1/5
 - Model = Simple Logistic Regression 	 f1-score = 0.5093	 precision = 0.8421 	 recall = 0.36502.0
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.7295	 precision = 0.7439 	 recall = 0.7158
 - Model = Deep Model (8 Dense Layers) 	 f1-score = 0.7538	 precision = 0.7647 	 recall = 0.7433
2. Eon 2/5
 - Model = Simple Logistic Regression 	 f1-score = 0.5780	 precision = 0.8399 	 recall = 0.44062.0
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.7526	 precision = 0.7822 	 recall = 0.7252
 - Model = Deep Model (8 Dense Layers) 	 f1-score = 0.7432	 precision = 0.7884 	 recall = 0.7030
3. Eon 3/5
 - Model = Simple Logistic Regression 	 f1-score = 0.6246	 precision = 0.8295 	 recall = 0.50092.0
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.7354	 precision = 0.8096 	 recall = 0.6737
 - Model = Deep Model (8 Dense Layers) 	 f1-score = 0.7558	 precision = 0.7734 	 recall = 0.7392
4. Eon 4/5
 - Model = Simple Logistic Regression 	 f1-score = 0.6462	 precision = 0.8281 	 recall = 0.52992.0
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.7599	 precision = 0.7807 	 recall = 0.7403
 - Model = Deep Model (8 Dense Layers) 	 f1-score = 0.7572	 precision = 0.7740 	 recall = 0.7412
5. Eon 5/5
 - Model = Simple Logistic Regression 	 f1-score = 0.6693	 precision = 0.8202 	 recall = 0.56542.0
 - Model = Deep Model (5 Dense Layers) 	 f1-score = 0.7476	 precision = 0.8020 	 recall = 0.7001
 - Model = Deep Model (8 Dense Layers) 	 f1-score = 0.7596	 precision = 0.7788 	 recall = 0.7414